src/main/java/jenkins/plugins/zulip/ZulipNotifier.java

```diff
@@ -74,73 +83,170 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnul
+        Map<String, StringBuilder> personalChangeStrings = new HashMap<String, StringBuilder>();
+        if (changes != null && !changes.isEmpty()) {
+            for (CommitInfo change : changes) {
+                if (!ZulipUtil.isValueSet(change.email)) {
+                    continue;
+                }
+
+                StringBuilder commitMsgs = null;
+                if (personalChangeStrings.containsKey(change.email)) {
+                    commitMsgs = personalChangeStrings.get(change.email);
+                } else {
+                    commitMsgs = new StringBuilder();
+                    commitMsgs.append("Changes since the last build:\n");
+                }
+                commitMsgs.append("\n* ").append(change.message);
+
+                personalChangeStrings.put(change.email, commitMsgs);
+            }
+        }
+        return personalChangeStrings;
+    }
+
+    private String buildCommonChangeString(List<CommitInfo> changes) {
         StringBuilder changeString = new StringBuilder();
+        if (changes == null) {
+            changeString.append("Could not determine changes since the last build.");
+        } else if (!changes.isEmpty()) {
+            // If there seems to be a commit message at all, try to list all the changes.
+            changeString.append("Changes since the last build:\n");
+            for (CommitInfo change : changes) {
+                if (ZulipUtil.isValueSet(change.email)) {
+                    changeString.append("\n* ");
+                    changeString.append(hundsonUrlMesssage("", "`" + change.author + "`", "mailto:" + change.email));
+                    changeString.append(" ");
+                } else {
+                    changeString.append("\n* `").append(change.author).append("` ");
+                }
+                changeString.append(change.message);
+            }
+        }
+        return changeString.toString();
+    }
+
+    private List<CommitInfo> getChangeSet(Run<?, ?> build) {
+        List<CommitInfo> result = new ArrayList<CommitInfo>();
         RunChangeSetWrapper wrapper = new RunChangeSetWrapper(build);
         if (!wrapper.hasChangeSetComputed()) {
-            changeString.append("Could not determine changes since last build.");
+            return null;
         } else if (wrapper.hasChangeSet()) {
-            // If there seems to be a commit message at all, try to list all the changes.
-            changeString.append("Changes since last build:\n");
             for (ChangeLogSet<? extends ChangeLogSet.Entry> changeLogSet : wrapper.getChangeSets()) {
                 for (ChangeLogSet.Entry e : changeLogSet) {
                     String commitMsg = e.getMsg().trim();
                     if (commitMsg.length() > 47) {
                         commitMsg = commitMsg.substring(0, 46) + "...";
                     }
+                    // GitChangeSet uses email as author's Id.
+                    String email = "";
```

Comment:
Did you manage to test this on both Git and Mercurial?

Can you create an utility functions that will extract email / display name from the commit info? The current implementation is hard to read. It will also be easier to write thorough tests for the detection.